Lib/re.py

```diff
@@ -274,8 +274,9 @@ def escape(pattern):
 _MAXCACHE = 512
 def _compile(pattern, flags):
     # internal: compile pattern
+    cache_key = (type(pattern), pattern, type(flags), flags)
```

Comment:
Putting type(flags) in the cache key prevents to merge flags=2 with flags=re.IGNORECASE.

I suggest to first focus on correctness. I let @methane handles the performance part with his PR #3862 :-)

But I propose to backport my fix to 2.7 and 3.6 without the optimization, so a minor performance regression for Python 3.6.